rfcs/90-re-support-initialization-via-config-string.md

```diff
@@ -0,0 +1,139 @@
+
+There are two reasons why config strings are removed, as mentioned in [GSP-13]:
+
+1. Usability: If an application uses [go-storage]'s config string, then it either exposes the format of config string to end user directly (additional burden for end users) or it has to manipulate its config into our format (additional burden for application developers). And if the application wants to combine its other config into a config string, this is also inconvenient.
+2. Implementation: We didn't find a good way to parse pairs from strings easily and safely.
+
+So we let the users of [go-storage] construct pairs directly using `WithXxx()`. 
+
+But passing string config is indeed more convenient, if not much. And actually parsing string into pairs is not that difficult. We can use a registry mechanism like [GSP-48]'s implementation.
+
+## Proposal
+
+So I propose to support service init from config string:
+
+We add the following APIs:
+```go
+func NewServicerFromString(config string) (types.Servicer, error) {}
+func NewStoragerFromString(config string) (types.Storager, error) {}
+```
+
+### Format
+
+The format of the config string is (optional parts marked by squared brackets):
+
+`<type>://[<name>][<work_dir>][?key1=value1&...&keyN=valueN]`
+
+- name: storage name, e.g., bucket name. MUST NOT contain /
+- work_dir: For object storage, it is prefix; for fs, it is directory path. MUST start with / for every storage services.
+
+So a valid config string could be:
+
+- `qingstor://bucket_name`
+- `qingstor://bucket_name/prefix`
+- `qingstor://bucket_name/prefix?credential=hmac:xxxx:xxxx&endpoint=https:qingstor.com:443`
+- `fs:///tmp`
+
+#### Parseable Value Types
+
+- `string`
+- `bool`
+- `int`
+- `int64`
+- `[]byte`: input should be base-64 encoded
+
+For types like `credential` and `endpoint`, they are parsed in specific service's `new` functions, and treated as plain string here.
+
+
+In the future, we may consider supporting complex types like `struct` (`DefaultStoragePairs`).
+
+#### Escaping?
+
+We don't require (and don't allow) users to escape the options. However, `&` SHOULD NOT be used since we use it as the delimiter. Other characters are OK (`/`, `=`, `+`, ...).
+
+In the future, we may consider escaping and enabling `&`.
+
+### Implementation
+
+`New*FromString` will first split config string into `(ty string, m map[string]string)`, and then parse it into `ps []Pairs`, and finally call `New*(ty, ps)`.
+To support this, we have to know that a name is a pair (global or service), and its type, so we implement:
+
+#### Pair Registry
+
+We register the types of global pairs and service pairs.
+
+From [go-storage] side:
+
+```go
+func RegisterServiceSchema(ty string, m map[string]string) {}
+```
+
+From services side, we can generate following code:
+
+```go
+var pairMap = map[string]string{
+	"xxxxx": "xxxxx",
+}
+
+func init() {
+	// ...
+	pairs.RegisterServiceSchema("<type>", pairMap)
```

Comment:
How about put this function in `services` like `services.RegisterSchema`?

Reply:
Oh, actually it is in `services`. not updated here